File: src/endpoint/Bc.Endpoint/CommentRegistration.cs
using System;
using System.Threading.Tasks;
using Bc.Contracts.Externals.Endpoint.CommentRegistration.Events;
using Bc.Contracts.Internals.Endpoint.CommentRegistration.Commands;
using Bc.Contracts.Internals.Endpoint.CommentRegistration.Logic;
using Bc.Contracts.Internals.Endpoint.GitHubPullRequestCreation.Messages;
using NServiceBus;

namespace Bc.Endpoint
{
    public class CommentRegistrationPolicy :
        Saga<CommentRegistrationPolicy.PolicyData>,
        IAmStartedByMessages<RegisterComment>,
        IHandleMessages<ResponseCreateGitHubPullRequest>
    {
        private readonly ICommentRegistrationPolicyLogic logic;

        public CommentRegistrationPolicy(ICommentRegistrationPolicyLogic logic)
        {
            this.logic = logic;
        }

        public Task Handle(RegisterComment message, IMessageHandlerContext context)
        {
            this.Data.UserName = message.UserName;
            this.Data.UserWebsite = message.UserWebsite;
            this.Data.UserComment = message.UserComment;
            this.Data.ArticleFileName = message.ArticleFileName;
            this.Data.CommentAddedDate = message.CommentAddedDate;

            var formatUserName = this.logic.FormatUserName(this.Data.UserName, this.Data.UserWebsite);
            var formatUserComment =
                this.logic.FormatUserComment(formatUserName, this.Data.UserComment, this.Data.CommentAddedDate);

            return context.Send(new RequestCreateGitHubPullRequest(
                message.CommentId,
                this.Data.ArticleFileName,
                formatUserComment,
                this.Data.CommentAddedDate));
        }

        public Task Handle(ResponseCreateGitHubPullRequest message, IMessageHandlerContext context)
        {
            this.MarkAsComplete();
            return context.Publish(new CommentRegistered(this.Data.CommentId, message.PullRequestUri));
        }

        protected override void ConfigureHowToFindSaga(SagaPropertyMapper<PolicyData> mapper)
        {
            mapper.ConfigureMapping<RegisterComment>(message => message.CommentId)
                  .ToSaga(data => data.CommentId);

            mapper.ConfigureMapping<ResponseCreateGitHubPullRequest>(message => message.CommentId)
                .ToSaga(data => data.CommentId);
        }

        public class PolicyData : ContainSagaData
        {
            public Guid CommentId { get; set; }

            public string UserName { get; set; }

            public string UserWebsite { get; set; }

            public string UserComment { get; set; }

            public string ArticleFileName { get; set; }

            public DateTime CommentAddedDate { get; set; }
        }
    }
}

File: src/endpoint/Bc.Endpoint/CommentRegistration.fs
module Bc.CommentRegistration

open System
open System.Threading.Tasks
open Bc.Contracts.Externals.Endpoint.CommentRegistration.Events
open Bc.Contracts.Internals.Endpoint.CommentRegistration.Commands
open Bc.Contracts.Internals.Endpoint.GitHubPullRequestCreation.Messages
open NServiceBus
open NServiceBus.Persistence.Sql

[<RequireQualifiedAccess>]
module Logic =

    let formatUserName userName userWebsite =
        match userWebsite with
        | null -> sprintf "**%s**" userName
        | "" -> sprintf "**%s**" userName
        | _ -> sprintf "[%s](%s)" userName userWebsite

    let formatUserComment userName userComment (commentAddedDate: DateTime) =
        sprintf "begin-%s-%s-%s UTC %s" userName userComment (commentAddedDate.ToString("yyyy-MM-dd HH:mm")) Environment.NewLine

type PolicyData() =
    inherit ContainSagaData()

    member val CommentId = Guid.Empty with get, set
    member val UserName = "" with get, set
    member val UserWebsite = "" with get, set
    member val UserComment = "" with get, set
    member val ArticleFileName = "" with get, set
    member val CommentAddedDate = DateTime.MinValue with get, set

[<SqlSaga(nameof Unchecked.defaultof<PolicyData>.CommentId)>]
type CommentRegistrationPolicy() =
    inherit Saga<PolicyData>()
        override this.ConfigureHowToFindSaga(mapper: SagaPropertyMapper<PolicyData>) =
            mapper.MapSaga(fun saga -> saga.CommentId :> obj)
                .ToMessage<RegisterComment>(fun message -> message.CommentId :> obj)
                .ToMessage<ResponseCreateGitHubPullRequest>(fun message -> message.CommentId :> obj) |> ignore

    interface IAmStartedByMessages<RegisterComment> with
        member this.Handle(message, context) =
            this.Data.UserName <- message.UserName
            this.Data.UserWebsite <- message.UserWebsite
            this.Data.UserComment <- message.UserComment
            this.Data.ArticleFileName <- message.ArticleFileName
            this.Data.CommentAddedDate <- message.CommentAddedDate

            let formatUserName = Logic.formatUserName this.Data.UserName this.Data.UserWebsite
            let formatUserComment = Logic.formatUserComment formatUserName this.Data.UserComment this.Data.CommentAddedDate

            context.Send(RequestCreateGitHubPullRequest(
                            message.CommentId,
                            this.Data.ArticleFileName,
                            formatUserComment,
                            this.Data.CommentAddedDate))

    interface IHandleMessages<ResponseCreateGitHubPullRequest> with
        member this.Handle(message, context) =
            this.MarkAsComplete()
            context.Publish(CommentRegistered(this.Data.CommentId, message.PullRequestUri))